### ISSUE TITLE
Tsne has no transform

### ISSUE THREAD

**Issue by BingqingCheng** [BingqingCheng/ASAP#22]
@Ryan-Rhys 
was using tsne.py to plot atomic projections using 
tsne.py -fxyz ASAP-soapparam-smart.xyz -fmat SOAPPARAM-smart --prefix amorphous-carbon -colors avg_bond --plotatomic true

But it appears that TSNE do not have the transform method.

Log below
load xyz file:  ASAP-soapparam-smart.xyz , a total of  50 frames , a total of  6250 atoms
Use descriptor matrix with shape:  (50, 441)
Use atomic descriptor matrix with shape:  (6250, 441)
shape of the descriptor matrix:  (50, 441) number of descriptors:  (441,)
Shape of processed descriptor matrix after applying PCA is (50, 50)
Traceback (most recent call last):
  File "/home/bingqing/.local/bin/tsne.py", line 4, in <module>
    __import__('pkg_resources').run_script('asaplib==0.0.1', 'tsne.py')
  File "/home/bingqing/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 667, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/bingqing/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1471, in run_script
    exec(script_code, namespace, namespace)
  File "/home/bingqing/.local/lib/python3.6/site-packages/asaplib-0.0.1-py3.6.egg/EGG-INFO/scripts/tsne.py", line 221, in <module>
  File "/home/bingqing/.local/lib/python3.6/site-packages/asaplib-0.0.1-py3.6.egg/EGG-INFO/scripts/tsne.py", line 101, in main
AttributeError: 'TSNE' object has no attribute 'transform'

**Comment by Ryan-Rhys** [comment]
TLDR: t-SNE as originally defined isn't meant to be used on out-of-sample data.

**scikit-learn discussion:**
https://github.com/scikit-learn/scikit-learn/issues/5361

**Discussion by co-originator of t-SNE (last comment in FAQ)**
https://lvdmaaten.github.io/tsne/

**Introduction of parametric t-SNE that can be used on out-of-sample data**
https://lvdmaaten.github.io/publications/papers/AISTATS_2009.pdf

**Comment by BingqingCheng** [comment]
So we should remove such option in tsne.py

**Comment by Ryan-Rhys** [comment]
Ah I see. What's the use-case for the atomic descriptors? If it's as straightforward as viewing the projection of a different set of descriptors then the defunct transform method could be replaced with a fit_transform method.

**Comment by BingqingCheng** [comment]
So one wants to build a map using the descriptors of the whole structure (avg of atomic descriptors), and then project the atomic descriptor matrix onto this map (out-of-sample proj)

**Comment by gabor1** [comment]
isn’t there a way to do a much small/faster a subsequent tSNE to do the "projection" of the new data? 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by Ryan-Rhys** [comment]
@gabor1 there's a parametric version of t-SNE referenced above which does this:

https://lvdmaaten.github.io/publications/papers/AISTATS_2009.pdf

This could be implemented if there's a good reason to believe that t-SNE would have substantial usage.

The simplest solution would just be to apply UMAP instead of t-SNE. It supports out-of-sample projection and the API's transform method works in exactly the same way as (k)pca. It is also claimed (although this has later been debated) that UMAP offers several advantages over t-SNE:

https://www.nature.com/articles/nbt.4314

I don't know enough about the applications here to gauge whether they are sensitive to the t-SNE vs. UMAP debate.

**Comment by gabor1** [comment]
Very good point. I would be quite interested in trying both UMAP and tSNE on our kinds of problems to substantiate or otherwise these claims

The parametric thing brings in regression. I'm wondering if it's simpler (but more costly?) to do another tSNE optimisation where one draws from the distribution of distances between the new points and the old points. This is how sketch-map or MDS would solve this problem.

**Comment by BingqingCheng** [comment]
I'm a bit lost. So is there a simple way of having this transform for tsne? It's very interesting to compare it with UMAP.

**Comment by gabor1** [comment]
depends on what you mean by simple. I’m saying it’s another tSNE run to place new points down. but it’s worth doing

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/
